### PR TITLE
Hops Away implementation

### PR DESCRIPTION
**Pull request by DayleDrinkwater**
A basic implementation of Hops Away, closing #922 

Modifies SignalInfo, if hopsAway is more than 0 it replaces the RSSI with the number of hops.

hops

### CODE REVIEW

**Comment by prokrypt** [comment]
Nice! But it leaves an ambiguity: a node with RSSI/SNR displayed is either a neighbor or a packet from an older firmware node/repeater/router. Perhaps try to a way to differentiate between the two...

**Comment by DayleDrinkwater** [comment]
Hi @prokrypt , good point. I've now made some changes. It now marks RSSI on nodes with a (!) if `hopStart` is 0, but `hopLimit `isn't. this will differentiate between nodes which are direct (no (!)) and nodes which are running old FW (not sending `hopStart`).

by default it marks all nodes with (!) until it knows for certain, at which point it will remove the ! if the are direct, or show Hops Away. if they are running old FW or it doesn't know, it'll continue to mark them with a (!)
hops2

**Review by prokrypt — COMMENTED** [review]
Possible lifehack:
`if (packet.hopStart < packet.hopLimit) ...` (just a random musing. totally untested!)

**Comment by DayleDrinkwater** [comment]
I did think that after I commited, I'll try it later on

File: app/src/main/java/com/geeksville/mesh/NodeInfo.kt [hunk 1]
```diff
@@ -219,6 +220,8 @@ data class NodeInfo(
     var snr: Float = Float.MAX_VALUE,
     var rssi: Int = Int.MAX_VALUE,
     var lastHeard: Int = 0, // the last time we've seen this node in secs since 1970
+    @ColumnInfo(name="hopsAway", defaultValue = "-1")
+    var hopsAway: Int = -1,
```
Comment (andrekir) [inline app/src/main/java/com/geeksville/mesh/NodeInfo.kt]: 1. this need to be added at the end of the object, or it creates a breaking change for the API
2. there is no need to add **@ColumnInfo**

File: .idea/inspectionProfiles/Project_Default.xml [hunk 2]
```diff
@@ -2,15 +2,39 @@
...
+    </inspection_tool>
+    <inspection_tool class="PreviewNotSupportedInUnitTestFiles" enabled="true" level="ERROR" enabled_by_default="true">
+      <option name="composableFile" value="true" />
+      <option name="previewFile" value="true" />
+    </inspection_tool>
+    <inspection_tool class="PreviewPickerAnnotation" enabled="true" level="ERROR" enabled_by_default="true">
+      <option name="composableFile" value="true" />
       <option name="previewFile" value="true" />
     </inspection_tool>
   </profile>
```
Comment (andrekir) [inline .idea/inspectionProfiles/Project_Default.xml]: please do not check in files that are unrelated to this pull request

File: app/src/main/java/com/geeksville/mesh/service/MeshService.kt [hunk 3]
```diff
@@ -1020,11 +1020,20 @@ class MeshService : Service(), Logging {
                 it.snr = packet.rxSnr
                 it.rssi = packet.rxRssi
             }
-
+            //generate our own hopsAway, comparing hopStart to hopLimit.
+            when {
```
Comment (andrekir) [inline app/src/main/java/com/geeksville/mesh/service/MeshService.kt]: setting `hopsAway` needs to move inside the `updateNodeInfo()` block [above](https://github.com/meshtastic/Meshtastic-Android/blob/b69466f669ce360577267e9adc41a970abae0421/app/src/main/java/com/geeksville/mesh/service/MeshService.kt#L1014). extra calls will trigger unnecessary node change broadcasts for every packet received.

this should match how `hopsAway` is calculated in the firmware.

**Review by andrekir — CHANGES_REQUESTED** [review]

File: .idea/inspectionProfiles/Project_Default.xml (hunk 2 above)
Comment (DayleDrinkwater) [reply .idea/inspectionProfiles/Project_Default.xml]: Apologies, I must've missed that one, I've reverted it now

File: app/src/main/java/com/geeksville/mesh/service/MeshService.kt (hunk 3 above)
Comment (DayleDrinkwater) [reply app/src/main/java/com/geeksville/mesh/service/MeshService.kt]: I've now moved `hopsAway `in to the `updateNodeInfo()` block as required.

Regarding `hopsAway `calculation, this essentially does follow the Firmware implementation, it checks if `hopStart `is not 0, then `hopStart `- `hopLimit`.

The one thing it does extra from the firmware is it checks if `hopStart` is less than `hopLimit`, if it is, set `hopsAway `to -1. It then uses this to append `(!)` to the RSSI data, to signify that it's probably from a device with old Firmware. 

The firmware version of this implementation does not differentiate between a packet which is received direct, and a packet from a node with old firmware (both return 0). This can be confusing to the user, and may make them think all nodes are local to them.

Is this OK? If not I have no problem removing it.

File: app/src/main/java/com/geeksville/mesh/NodeInfo.kt (hunk 1 above)
Comment (DayleDrinkwater) [reply app/src/main/java/com/geeksville/mesh/NodeInfo.kt]: I've now moved hopsAway to the end.

I'm having issues removing `@ColumnInfo` though, when I remove that line it will not build, I receive the error `"New NOT NULL column'hopsAway' added with no default value specified. Please specify the default value using @ColumnInfo."`. I'm not exactly sure why it is required compared to other columns, but looking online it's related to versioning of the Room database.

File: app/src/main/java/com/geeksville/mesh/service/MeshService.kt (hunk 3 above)
Comment (andrekir) [reply app/src/main/java/com/geeksville/mesh/service/MeshService.kt]: keep it simple and align with the code in the firmware. `hopsAway` from the radio (removed in [`b69466f`](https://github.com/meshtastic/Meshtastic-Android/pull/966/commits/b69466f669ce360577267e9adc41a970abae0421#diff-3a6bc031c4bd5b2872f201d2b4b98ed404f169f548d64821749bd6144efaf854L1322)) must be used and any extra handling must be compatible.

keep in mind all this added complexity will be unnecessary in a few weeks once everyone upgrades to 2.3.0 or later. if there is room for improvement here or on the firmware implementation, address them in separate PRs.

Comment (DayleDrinkwater) [reply app/src/main/java/com/geeksville/mesh/service/MeshService.kt]: no problem, I've now removed the extras and `hopsAway` now follows the firmware implementation.

Comment (andrekir) [reply app/src/main/java/com/geeksville/mesh/service/MeshService.kt]: thanks

**Review by andrekir — APPROVED** [review]
